### PR TITLE
External entity references

### PR DESCRIPTION
**Pull request by Harri-H**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -37,7 +37,8 @@
 	"dependencies": {
 		"lodash": "4.17.10",
 		"marc-record-js": "git+https://github.com/natlibfi-arlehiko/marc-record-js.git",
-		"ramda": "0.25.0"
+		"node-fetch": "^2.1.2",
+		"xml2js": "^0.4.19"
```
Comment (natlibfi-arlehiko) [inline package.json]: Change this to `0.x.x`

File: src/external-entity-references-melinda.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+ * @licend  The above is the entire license notice
+ * for the JavaScript code in this file.
+ *
+ */
+
+import {parseString} from 'xml2js';
+import fetch from 'node-fetch';
+
+export default async function ({endpoint, prefixPattern, fields}) {
+	if (prefixPattern instanceof RegExp && typeof fields === 'object') {
```
Comment (natlibfi-arlehiko) [inline src/external-entity-references-melinda.js]: Also check that endpoint is a string

File: src/external-entity-references-melinda.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+							prev.push(item);
+						}
+						return prev;
+					});
+				}
+			});
+			return prev;
+		}, []);
+
+		// Matching prefixPatter is removed from object value field.
```
Comment (natlibfi-arlehiko) [inline src/external-entity-references-melinda.js]: Typo: `prefixPatter` -> `prefixPattern`

File: src/external-entity-references-melinda.js [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+		matchingTags.forEach(obj => {
+			if (prefixPattern.test(obj.value)) {
+				obj.value = obj.value.replace(prefixPattern, '');
+				removedPrefixes.push(obj);
+			}
+		});
+		// If matching prefixPatterns found make an API call
+		if (removedPrefixes.length > 0) {
+			validationResult = validateMatcingTags(removedPrefixes);
```
Comment (natlibfi-arlehiko) [inline src/external-entity-references-melinda.js]: Typo: `validateMatcingTags` -> `validateMatchingTags`

File: src/external-entity-references-melinda.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+		// Matching prefixPatter is removed from object value field.
+		matchingTags.forEach(obj => {
+			if (prefixPattern.test(obj.value)) {
+				obj.value = obj.value.replace(prefixPattern, '');
+				removedPrefixes.push(obj);
+			}
+		});
+		// If matching prefixPatterns found make an API call
+		if (removedPrefixes.length > 0) {
```
Comment (natlibfi-arlehiko) [inline src/external-entity-references-melinda.js]: If there are no values to function returns `{valid: false}` because `validationResult` is initalized to `false`.

The validation should fail only when any of the values is not resolvable. The validation succeeds if there are no values to check.

File: test/external-entity-references-melinda.spec.js [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+								code: 'w',
+								value: '(FI-MELINDA)123456'
+							}
+						]
+					}
+				]
+			});
+			const result = await validator.validate(record);
+
+			expect(result).to.eql({valid: false});
```
Comment (natlibfi-arlehiko) [inline test/external-entity-references-melinda.spec.js]: Should be `{valid: true}` since finding to values is ok.

File: test/external-entity-references-melinda.spec.js [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+const endpoint = 'http://melinda.kansalliskirjasto.fi:210/fin01';
+const queryParam = '?operation=searchRetrieve&maximumRecords=2&version=1&query=rec.id=';
+
+const prefixPattern = /^\(FOOBAR\)/;
+const fields = {
+	773: ['w'],
+	833: ['w', 'p']
+};
+
+describe('external-entity-references-melinda', () => {
```
Comment (natlibfi-arlehiko) [inline test/external-entity-references-melinda.spec.js]: Add a test for a case where values to resolve are found but are not resolvable (The SRU API returns zero or more than one records).

**Comment by natlibfi-arlehiko** [comment]
Also return validation messags when validation fails:
```js
{
  valid:false,
  messages: [
    'Field 773$w with value \'(FOOBAR)4000\' is not resolvable'
  ]
}
```

**Comment by natlibfi-arlehiko** [comment]
Validator name should be `resolvable-ext-references-melinda`

File: package.json (hunk 1 above)
Comment (Harri-H) [reply package.json]: fixed

File: src/external-entity-references-melinda.js (hunk 2 above)
Comment (Harri-H) [reply src/external-entity-references-melinda.js]: fixed

File: src/external-entity-references-melinda.js (hunk 4 above)
Comment (Harri-H) [reply src/external-entity-references-melinda.js]: fixed

File: src/external-entity-references-melinda.js (hunk 3 above)
Comment (Harri-H) [reply src/external-entity-references-melinda.js]: fixed

File: src/external-entity-references-melinda.js (hunk 5 above)
Comment (Harri-H) [reply src/external-entity-references-melinda.js]: fixed

File: test/external-entity-references-melinda.spec.js (hunk 6 above)
Comment (Harri-H) [reply test/external-entity-references-melinda.spec.js]: fixed

File: package.json [hunk 8]
```diff
@@ -38,7 +38,7 @@
 		"lodash": "4.17.10",
 		"marc-record-js": "git+https://github.com/natlibfi-arlehiko/marc-record-js.git",
 		"node-fetch": "^2.1.2",
-		"xml2js": "^0.4.19"
+		"xml2js": "^0.x.x"
```
Comment (natlibfi-arlehiko) [inline package.json]: Remove the caret -> `0.x.x`

File: test/external-entity-references-melinda.spec.js (hunk 7 above)
Comment (Harri-H) [reply test/external-entity-references-melinda.spec.js]: added

File: test/fixtures/external-entity-references.js [hunk 9]
```diff
@@ -0,0 +1,323 @@
+/**
```
Comment (natlibfi-arlehiko) [inline test/fixtures/external-entity-references.js]: Change filename to **resolvable-ext-references-melinda.js**

Comment (Harri-H) [reply test/fixtures/external-entity-references.js]: fixed
